browser/src/Renderer/WebGLRenderer/TextRenderer/TextRenderer.ts

```diff
@@ -285,43 +285,54 @@ export class WebGlTextRenderer {
         const pixelRatioAdaptedGlyphOverlap = this._glyphOverlapInPixels * this._devicePixelRatio
 
         let glyphCount = 0
-        let y = 0
-
-        // TODO refactor this to not be as reliant on mutations
         for (let rowIndex = 0; rowIndex < rowCount; rowIndex++) {
-            let x = 0
+            const cellGroups = groupCells(columnCount, rowIndex, getCell)
+
+            cellGroups.forEach(cellGroup => {
+                const { startColumnIndex, characters, bold, italic, foregroundColor } = cellGroup
 
-            for (let columnIndex = 0; columnIndex < columnCount; columnIndex++) {
-                const cell = getCell(columnIndex, rowIndex)
-                const char = cell.character
-                if (!isWhiteSpace(char)) {
+                const ligatureGroups = this._ligatureGrouper.getLigatureGroups(characters)
+
+                let offsetWithinCellGroup = 0
+                ligatureGroups.forEach(ligatureGroup => {
+                    const columnIndex = startColumnIndex + offsetWithinCellGroup
+                    const x = pixelRatioAdaptedFontWidth * columnIndex
```

Comment:
I'd be careful about this if you want to support non-monospace fonts.

Reply:
Thanks for the heads-up!

I don't think that non-monospace fonts will be relevant though, because we build the whole editor on the basis of Neovim, which is a grid-based editor that fundamentally works by using monospaced cells. If we decided to go away from this, not only will the keyboard navigation between lines potentially become hard to understand but also many existing features and plugins might not make sense anymore. That's why I think it's safe to assume that we can position all characters based on this monospaced grid without worrying about it.